tlvc/arch/i86/drivers/block/directfd.c

```diff
@@ -567,7 +571,11 @@ static void setup_DMA(void)
     } else if (dma_addr >= LAST_DMA_ADDR) {		/* Not likely */
 	dma_addr = _MK_LINADDR(kernel_ds, tmp_floppy_area); /* use bounce buffer */
 	if (command == FD_WRITE) {
+#ifdef CONFIG_FS_XMS_BUFFER
+	    xms_fmemcpyw(tmp_floppy_area, kernel_ds, CURRENT->rq_buffer, CURRENT->rq_seg, BLOCK_SIZE/2);
+#else
 	    fmemcpyw(tmp_floppy_area, kernel_ds, CURRENT->rq_buffer, CURRENT->rq_seg, BLOCK_SIZE/2);
+#endif
```

Comment:
An ifdef for `CONFIG_FS_XMS_BUFFER` isn't required here (or below) to differentiate between copying to/from XMS (using `xms_fmemcpy`) and `fmemcpyw` (to/from main memory) in any block driver. It is all handled automatically within `xms_fmemcpyw`, which can tell the difference between an XMS and main memory address, and will call `fmemcpy` if needed.

This is because the all buffer segment descriptors (including `CURRENT->rq_seg`) are declared of type `ramdesc_t` (an unsigned long), rather than `seg_t` (an unsigned short). A `ramdesc_t` (ram descriptor) is a 32-bit physical memory address when the upper 16 bits are nonzero, and an 8086 segment descriptor when the upper bits are 0. This format allows for passing a single address for buffers without having to specify XMS linear addresses or segment values separately.

A caveat of this is that when CONFIG_FS_XMS_BUFFER is defined, one can't assume that the lower 16 bits of a `CURRENT->rq_seg` is an 8086 segment value and shift left by 4 to obtain a physical address. The `ramdesc_t` values are interpreted within `xms_fmemcpyw` and the appropriate routine called.